llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,455 @@
+             } dbgs()
+             << "\n";);
+
+  return PhysRegs;
+}
+
+static bool isUndefReserved(
+    const MachineOperand &Def, const MachineInstr &MI,
+    const MachineBasicBlock &MBB,
+    std::map<unsigned, SmallVector<const MachineInstr *, 4>> &UndefPartReuse,
+    const VirtRegMap &VRM) {
+  unsigned ID = Def.getReg().id();
+  if (Def.isUndef()) {
+    if (UndefPartReuse.find(ID) == UndefPartReuse.end()) {
+      UndefPartReuse[ID] = SmallVector<const MachineInstr *, 4>();
+    }
+  }
+
+  if (UndefPartReuse.find(ID) == UndefPartReuse.end())
+    return false;
+
+  UndefPartReuse[ID].push_back(&MI);
+
+  // Undef Reg is still being filled, too early to tell, if it should be removed
+  if (Def.isUndef() && isNotLastDef(Def, MI, MBB))
+    return true;
+
+  // find if another SubRegister has already been assigned by a previous copy
+  auto CurrentPhysReg = VRM.getPhys(Def.getReg());
+  for (const auto &Reg : UndefPartReuse) {
+    if (Reg.first == ID)
+      continue;
+
+    if (VRM.getPhys(Reg.first) == CurrentPhysReg &&
+        reuseSubRegs(UndefPartReuse[ID], UndefPartReuse[Reg.first]))
+      return true;
+  }
+  return false;
+}
+
+static bool isNotLastDef(const MachineOperand &Def,
+                         const MachineInstr &DefInstr,
+                         const MachineBasicBlock &MBB) {
+  const Register Reg = Def.getReg();
+  bool FoundDef = false;
+  for (const MachineInstr &MI : MBB) {
+    if (!MI.definesRegister(Reg))
+      continue;
+    if (MI.isIdenticalTo(DefInstr)) {
+      FoundDef = true;
+      continue;
+    }
+    if (FoundDef)
+      return true;
+  }
+  return false;
+}
+
+static bool reuseSubRegs(SmallVector<const MachineInstr *, 4> &OrigInstr,
+                         SmallVector<const MachineInstr *, 4> &OtherInstr) {
+  for (auto [InstrA, InstrB] : zip(OrigInstr, OtherInstr)) {
+    if (!InstrA->isCopyLike())
+      continue;
+    if (!InstrB->isCopyLike())
+      continue;
+
+    for (auto [OpA, OpB] : zip(InstrA->uses(), InstrB->uses())) {
+      if (!OpA.isReg() || !OpB.isReg())
+        continue;
+      if (OpA.getReg() != OpB.getReg())
+        continue;
+      if (OpA.getSubReg() != OpB.getSubReg())
+        continue;
+
+      if (containEqualSubRegDefs(InstrA, InstrB))
+        return true;
+    }
+  }
+  return false;
+}
+
+static bool containEqualSubRegDefs(const MachineInstr *InstrA,
+                                   const MachineInstr *InstrB) {
+  for (auto [defA, defB] : zip(InstrA->defs(), InstrB->defs())) {
+    if (defA.getSubReg() == defB.getSubReg())
+      return true;
+  }
+  return false;
+}
+
+bool isCopyElimination(const MachineInstr &MI, const VirtRegMap &VRM) {
+  if (!MI.isCopy())
+    return false;
+
+  // if src and target are equal, this copy will be eliminated, and
+  // therefore should not renamed or even registered
+  assert(MI.getNumOperands() == 2 &&
+         "Copy MI does not have 2 operands, Naive Copy Elimination will "
+         "not work! Please fix.");
```

Comment:
I think we can remove that assert. Verifying properties like this is the job of the `MachineVerifier`

Reply:
removed